lib/logstash/outputs/elasticsearch_java/protocol.rb

```diff
@@ -0,0 +1,248 @@
+require "cabin"
+require "base64"
+require 'logstash-output-elasticsearch_java_jars.rb'
+require 'logstash/outputs/elasticsearch_java'
+
+module LogStash
+  module Outputs
+    module ElasticSearchJavaPlugins
+      module Protocols
+        class Base
+          private
+          def initialize(options={})
+            # host(s), port, cluster
+            @logger = Cabin::Channel.get
+          end
+
+          def client
+            return @client if @client
+            @client = build_client(@options)
+            return @client
+          end # def client
+
+
+          def template_install(name, template, force=false)
+            if template_exists?(name) && !force
+              @logger.debug("Found existing Elasticsearch template. Skipping template management", :name => name)
+              return
+            end
+            template_put(name, template)
+          end
+
+          # Do a bulk request with the given actions.
+          #
+          # 'actions' is expected to be an array of bulk requests as string json
+          # values.
+          #
+          # Each 'action' becomes a single line in the bulk api call. For more
+          # details on the format of each.
+          def bulk(actions)
+            raise NotImplemented, "You must implement this yourself"
+            # bulk([
+            # '{ "index" : { "_index" : "test", "_type" : "type1", "_id" : "1" } }',
+            # '{ "field1" : "value1" }'
+            #])
+          end
+
+          public(:initialize, :template_install)
+        end
+
+        class NodeClient < Base
+          private
+
+          DEFAULT_OPTIONS = {
+            :port => 9300,
+          }
+
+          def initialize(options={})
+            super
+            require "java"
+            @options = DEFAULT_OPTIONS.merge(options)
+            setup(@options)
+            @client = client
+          end # def initialize
+
+          def settings
+            return @settings
+          end
+
+          def setup(options={})
+            @settings = org.elasticsearch.common.settings.Settings.settingsBuilder()
```

Comment:
This is, but it's pretty much unchanged from previous work. I think we should save that for a future PR.

This plugin needs a lot of work :/